vpr/src/util/vpr_utils.cpp

```diff
@@ -2267,27 +2279,79 @@ int get_sub_tile_physical_pin(int sub_tile_index,
     return result->second.pin;
 }
 
+int get_logical_block_physical_sub_tile_index(t_physical_tile_type_ptr physical_tile,
+                                              t_logical_block_type_ptr logical_block) {
+    int sub_tile_index = OPEN;
+    for (const auto& sub_tile : physical_tile->sub_tiles) {
+        auto eq_sites = sub_tile.equivalent_sites;
+        auto it = std::find(eq_sites.begin(), eq_sites.end(), logical_block);
+        if (it != eq_sites.end()) {
+            sub_tile_index = sub_tile.index;
+        }
+    }
+
+    if (sub_tile_index == OPEN) {
+        VPR_ERROR(VPR_ERROR_OTHER,
+                  "Found no instances of logical block type '%s' within physical tile type '%s'. ",
+                  logical_block->name, physical_tile->name);
+    }
+
+    return sub_tile_index;
+}
+
 int get_physical_pin(t_physical_tile_type_ptr physical_tile,
                      t_logical_block_type_ptr logical_block,
                      int pin) {
+    int sub_tile_index = get_logical_block_physical_sub_tile_index(physical_tile, logical_block);
+
+    if (sub_tile_index == OPEN) {
+        VPR_ERROR(VPR_ERROR_OTHER,
+                  "Couldn't find the corresponding physical tile type pin of the logical block type pin %d.",
+                  pin);
+    }
+
+    int sub_tile_physical_pin = get_sub_tile_physical_pin(sub_tile_index, physical_tile, logical_block, pin);
+    return physical_tile->sub_tiles[sub_tile_index].sub_tile_to_tile_pin_indices[sub_tile_physical_pin];
+}
+
+int get_logical_block_physical_sub_tile_index(t_physical_tile_type_ptr physical_tile,
+                                              t_logical_block_type_ptr logical_block,
+                                              int sub_tile_capacity) {
     int sub_tile_index = OPEN;
     for (const auto& sub_tile : physical_tile->sub_tiles) {
         auto eq_sites = sub_tile.equivalent_sites;
         auto it = std::find(eq_sites.begin(), eq_sites.end(), logical_block);
-        if (it != eq_sites.end()) {
+        if (it != eq_sites.end()
+            && (sub_tile.capacity.is_in_range(sub_tile_capacity))) {
             sub_tile_index = sub_tile.index;
+            break;
         }
     }
 
     if (sub_tile_index == OPEN) {
         VPR_ERROR(VPR_ERROR_OTHER,
-                  "Found no instances of logical block type '%s' within physical tile type '%s'. "
+                  "Found no instances of logical block type '%s' within physical tile type '%s'. ",
+                  logical_block->name, physical_tile->name);
+    }
+
+    return sub_tile_index;
+}
+
+int get_physical_pin(t_physical_tile_type_ptr physical_tile,
+                     t_logical_block_type_ptr logical_block,
+                     int sub_tile_capacity,
+                     int pin) {
+    int sub_tile_index = get_logical_block_physical_sub_tile_index(physical_tile, logical_block, sub_tile_capacity);
```

Comment:
You should be able to get the sub_tile_index with the already present  function `get_sub_tile_index`: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/37095c742471af2e2c8d90faf10ab13156a47942/vpr/src/util/vpr_utils.cpp#L721-L745

This requires the ClusterBlockId to retrieve the placed location, but I think it is worth using this instead of adding a new overloaded `get_logical_block_physical_sub_tile_index` function.

Reply:
Ok, so, if you don't have at disposal the ClusterBlockID when calling this function, you cannot make use of the already existing helper function.

I think that the best way forward (probably in a follow-up PR) is to encapsulate these helper functions, that need to rely only on the device information, within the physical tile type data model, as previously discussed. For the time being, having overloaded functions does not seem to be a big issue.

Once I'll get to add some more description on how this pin mapping works, I will see if I can correctly encapsulate the functions.

Reply:
No worries. I think it would be helpful to clarify in the comments which APIs rely only on the device information and which need mapping information. That will better guide future developers when reworking these APIs. At least, in this pull request, I can add comments to clarify these points. We just need to agree on whether these functions should require only device information or not.